rfcs/20230420-kernel-cache/README.md

```diff
@@ -0,0 +1,292 @@
+
+Cons:
+* There is no guarantee that all kernels in the kernel cache are used by all the
+primitives in the primitive cache. It might happen those primitives in the
+primitive cache has unique kernels and the kernel cache may have unique kernels
+too. This means that memory consumption can be higher that it is now
+
+##### Option 2: Provide API to Control Kernel Cache Capacity
+
+This option suggests adding an API to control kernel cache capacity similar to
+the one for the primitive cache.
+
+Pros:
+* Provides more control over the kernel cache and therefore more flexibility to
+the users. Those users who heavily rely on dynamic shapes can balance between
+primitive cache capacity and the kernel cache capacity
+* User can fully control memory consumption
+
+Cons:
+* Introducing yet another *cache capacity* definition that should be specified.
+Also, we will have to explain to the users the difference between kernel cache and primitive cache capacities as well as how they interoperate with each other
+
+###### C API
+
+```c
+dnnl_status_t dnnl_get_kernel_cache_capacity(int *capacity);
+dnnl_status_t dnnl_set_kernel_cache_capacity(int capacity);
+```
+
+###### C++ API
+
+```cpp
+int get_primitive_cache_capacity();
+void set_primitive_cache_capacity(int capacity);
+```
+
+###### Environment Variable
+
+`ONEDNN_PRIMITIVE_CACHE_CAPACITY=<number>`, where `number` is the kernel cache
+capacity. The default value is 1024. When `number` is 0 the kernel cache is
+disabled.
+
+###### Build Time Control
+
+The kernel cache can be enabled or disabled at the build time using a CMake option
+`ONEDNN_ENABLE_KERNEL_CACHE`.
+
+##### Options Comparison
+
+Both options assume that there will be a reasonable default capacity for the
+kernel cache. Also, both options allow users to control the caches capacity
+therefore, the only difference between the options is in the flexibility.
+The second option is in fact the first one with advanced controlling
+capabilities.
+
+Given that the first option will address the problem with dynamic shapes
+related creation overhead the proposal is to start with the it and add the
+API from the option two if we see any need in it.
+
+### Option 2: Extending Runtime Dimensions Support
+
+This option suggests extending the existing runtime dimensions support for
+relevant primitives.
+
+Pros:
+* Using the existing mechanism, the primitive cache will automatically work
+for primitives created with runtime dimensions
+* Unlike the kernel cache, with this approach the primitive and primitive descriptor
+is not re-created therefore there is no related overhead
+
+Cons:
+* The users have to modify their code to enable runtime dimensions
+* Some primitives will require to generate multiple kernels so that the
+primitives can be used with runtime dimensions. It may not be always feasible to
+generate a reasonable number of kernels
+* For some kernels it may not be possible to stay performant and support runtime
```

Comment:
Did this mean if user set wrong runtime dimensions, it would fall back to ref-path?

Reply:
Not necessarily.  This just means the exact nature of the performance gaps would be highly implementation specific and hard to predict. What is performant would largely depend on what optimized cases oneDNN developers identify as important.

Reply:
So, for kernel implementation, can we just set all dimensions as runtime dimension and let oneDNN to handle?
Can we get better performance based on this?

Reply:
Do we have performance data to show the performance drop with runtime dims?

Reply:
@guizili0,
> So, for kernel implementation, can we just set all dimensions as runtime dimension and let oneDNN to handle?

Yes, but there will be performance implications. For example, suppose the most performance kernel can only be used when all dimensions except one are non-runtime. Then the question is, what does the library do when the other dimensions are runtime ones? It will likely to go to the reference implementation that will have performance impact.

@gujinghui,
> Do we have performance data to show the performance drop with runtime dims?

I think we conducted some experiments in the past and concluded that the performance impact could be significant. 
@rjoursler @petercad may provide more details on the matter.

Reply:
@gujinghui, I don't know of any data, but the current actualized performance is not great as the current implementation just dispatches to the reference matmul implementation. As to a "theoretical best implementation", that is a complicated question which is highly dependent on resources tradeoffs.
Internally, oneDNN can generate many different kernel specializations for performing a primitive operation. To keep runtime dimensions maintainable, we would repurpose a set of these specializations to correspond to the primitive with runtime dimensions. Because of this, performance would largely correspond with how well the chosen specializations align with the measured workload. Basically, performance would be great so long as you are on the "good path", but could be very poor otherwise.  The only way around this is to allow all specializations to be generated. The upfront cost of this can be quite high, especially for primitives like matmul. To get around this, kernel creation would need to be performed lazily, at which point we have a structure similar to the kernel cache. Functionally, lazy kernel generation is very similar to the kernel cache, but there are some reusability/memory control tradeoffs as kernel reuse is only allowed within a primitive with runtime dimensions.

Reply:
@rjoursler If my understand right, you mean that oneDNN would map multi-kernel to one primitive to support this runtime dimension. So this would hurt the first iteration time as in first iteration, you would run multi-kernel jitting instead of one.